srv/salt/_modules/dg.py

```diff
@@ -0,0 +1,822 @@
+        return self.data_device_attrs.get("limit", 0)
+
+    @property
+    def inventory(self) -> dict:
+        """
+        Disks found in the inventory
+        """
+        return Inventory().disks
+
+    @property
+    def data_devices(self) -> list:
+        """ Filter for (bluestore) DATA devices
+        """
+        log.debug("Scanning for data devices")
+        return self._filter_devices(self.data_device_attrs)
+
+    @property
+    def wal_devices(self) -> list:
+        """ Filter for bluestore WAL devices
+        """
+        log.debug("Scanning for WAL devices")
+        return self._filter_devices(self.wal_device_attrs)
+
+    @property
+    def db_devices(self) -> list:
+        """ Filter for bluestore DB devices
+        """
+        log.debug("Scanning for db devices")
+        return self._filter_devices(self.db_device_attrs)
+
+    def _limit_reached(self, len_devices: int, disk_path: str) -> bool:
+        """ Check for the <limit> property and apply logic
+
+        If a limit is set in 'device_attrs' we have to stop adding
+        disks at some point.
+
+        If limit is set (>0) and len(devices) >= limit
+
+        :param int len_devices: Length of the already populated device set/list
+        :param str disk_path: The disk identifier (for logging purposes)
+        :return: True/False if the device should be added to the list of devices
+        :rtype: bool
+        """
+        if self.limit > 0 and len_devices >= self.limit:
+            log.info("Refuse to add {} due to limit policy of {}>".format(
+                disk_path, self.limit))
+            return True
+        return False
+
+    def _filter_devices(self, device_filter: dict) -> list:
+        """ Filters devices with applied filters
+
+        Iterates over all applied filter (there can be multiple):
+
+        size: 10G:50G
+        model: Fujitsu
+        rotational: 1
+
+        Question: ##############################
+        This currently acts as a OR gate. Should this be a AND gate?
+        Question: #############################
+
+        Iterates over all known disk and checks
+        for matches by using the matcher subclasses.
+
+        :param dict device_filter: Device filter as in description above
+        :return: Set of devices that matched the filter
+        :rtype set:
+        """
+        devices: Set = set()
+        for name, val in device_filter.items():
+            _filter = Filter(name=name, value=val)
+            for disk in self.inventory:
+                # continue criterias
+                if not _filter.is_matchable:
+                    continue
+
+                if not _filter.matcher.compare(self._reduce_inventory(disk)):
+                    continue
+
+                if not self._has_mandatory_idents(disk):
+                    continue
+
+                if self._limit_reached(len(devices), disk.get('path')):
+                    continue
+
+                devices.add(disk.get("path"))
```

Comment:
When a device is added it should be removed from the list of available devices.

Reply:
True. At least when talking about data_devices. This shouldn't apply for wal/db_devices.

I'll add that to my list and and push it with the next set of patches.